snow/consensus/snowman/topological.go

```diff
@@ -288,23 +308,28 @@ func (ts *Topological) RecordPoll(ctx context.Context, voteBag bag.Bag[ids.ID])
 		return nil
 	}
 
-	// Runtime = |live set| ; Space = Constant
+	// Runtime = 2 * |live set| ; Space = Constant
 	ts.preferredIDs.Clear()
+	maps.Clear(ts.preferredHeights)
 
 	ts.tail = preferred
 	startBlock := ts.blocks[ts.tail]
 
 	// Runtime = |live set| ; Space = Constant
 	// Traverse from the preferred ID to the last accepted ancestor.
 	for block := startBlock; !block.Accepted(); {
-		ts.preferredIDs.Add(block.blk.ID())
+		blkID := block.blk.ID()
+		ts.preferredIDs.Add(blkID)
+		ts.preferredHeights[block.blk.Height()] = blkID
 		block = ts.blocks[block.blk.Parent()]
 	}
 	// Traverse from the preferred ID to the preferred child until there are no
 	// children.
-	for block := startBlock; block.sb != nil; block = ts.blocks[ts.tail] {
+	for block := startBlock; block.sb != nil; {
 		ts.tail = block.sb.Preference()
 		ts.preferredIDs.Add(ts.tail)
+		block = ts.blocks[ts.tail]
+		ts.preferredHeights[block.blk.Height()] = ts.tail
 	}
```

Comment:
This is really the only subtle change - `block.sb` is only initialized after a child block is inserted. So, we are guaranteed that `block.blk` is non-nil after updating to `ts.blocks[ts.tail]`.

Reply:
Added as a comment